PowerEditor/src/WinControls/Grid/ShortcutMapper.cpp

```diff
@@ -124,123 +122,176 @@ void ShortcutMapper::initBabyGrid() {
+				if (isFilterValid(cshortcuts[i])) {
+					_babygrid.setText(cs_index, 1, cshortcuts[i].getName());
+					if (cshortcuts[i].isEnabled()) //avoid empty strings for better performance
+						_babygrid.setText(cs_index, 2, cshortcuts[i].toString().c_str());
+					_babygrid.setText(cs_index, 3, cshortcuts[i].getCategory());
+					if (isMarker)
+						isMarker = _babygrid.setMarker(false);
+					_shortcutIndex.push_back(i);
+					cs_index++;
+				}
 			}
+			_babygrid.setLineColNumber(cs_index - 1 , 3);
             ::EnableWindow(::GetDlgItem(_hSelf, IDM_BABYGRID_MODIFY), true);
             ::EnableWindow(::GetDlgItem(_hSelf, IDM_BABYGRID_CLEAR), true);
             ::EnableWindow(::GetDlgItem(_hSelf, IDM_BABYGRID_DELETE), false);
 			break; }
 		case STATE_MACRO: {
 			vector<MacroShortcut> & cshortcuts = nppParam->getMacroList();
+			cs_index = 1;
 			for(size_t i = 0; i < nbItems; ++i)
 			{
 				if (findKeyConflicts(nullptr, cshortcuts[i].getKeyCombo(), i))
 					isMarker = _babygrid.setMarker(true);
 
-				_babygrid.setText(i+1, 1, cshortcuts[i].getName());
-				if (cshortcuts[i].isEnabled()) //avoid empty strings for better performance
-					_babygrid.setText(i+1, 2, cshortcuts[i].toString().c_str());
-
-				if (isMarker)
-					isMarker = _babygrid.setMarker(false);
+				if (isFilterValid(cshortcuts[i])) {
+					_babygrid.setText(cs_index, 1, cshortcuts[i].getName());
+					if (cshortcuts[i].isEnabled()) //avoid empty strings for better performance
+						_babygrid.setText(cs_index, 2, cshortcuts[i].toString().c_str());
+	
+					if (isMarker)
+						isMarker = _babygrid.setMarker(false);
+					_shortcutIndex.push_back(i);
+					cs_index++;
+				}
 			}
+			_babygrid.setLineColNumber(cs_index - 1 , 2);
             bool shouldBeEnabled = nbItems > 0;
             ::EnableWindow(::GetDlgItem(_hSelf, IDM_BABYGRID_MODIFY), shouldBeEnabled);
             ::EnableWindow(::GetDlgItem(_hSelf, IDM_BABYGRID_CLEAR), shouldBeEnabled);
             ::EnableWindow(::GetDlgItem(_hSelf, IDM_BABYGRID_DELETE), shouldBeEnabled);
 			break; }
 		case STATE_USER: {
 			vector<UserCommand> & cshortcuts = nppParam->getUserCommandList();
+			cs_index = 1;
 			for(size_t i = 0; i < nbItems; ++i)
 			{
 				if (findKeyConflicts(nullptr, cshortcuts[i].getKeyCombo(), i))
 					isMarker = _babygrid.setMarker(true);
 
-				_babygrid.setText(i+1, 1, cshortcuts[i].getName());
-				if (cshortcuts[i].isEnabled()) //avoid empty strings for better performance
-					_babygrid.setText(i+1, 2, cshortcuts[i].toString().c_str());
-
-				if (isMarker)
-					isMarker = _babygrid.setMarker(false);
+				if (isFilterValid(cshortcuts[i])) {
+					_babygrid.setText(cs_index, 1, cshortcuts[i].getName());
+					if (cshortcuts[i].isEnabled()) //avoid empty strings for better performance
+						_babygrid.setText(cs_index, 2, cshortcuts[i].toString().c_str());
+	
+					if (isMarker)
+						isMarker = _babygrid.setMarker(false);
+					_shortcutIndex.push_back(i);
+					cs_index++;
+				}

```

Comment:
This is quite a lot of code duplication. Is there a way to avoid this?

Reply:
last year, I tried to factorize and avoid code duplication but that was too much code modifications. 
Sorry